File: packages/cli/src/commands-robots-watch/cursor.ts
// Persisted per-bead status cursor for the poll-daemon: the "what we last saw"
// half of the diff. { "<store>": { "<bead-id>": "<status>" } }.

import { existsSync, mkdirSync, readFileSync, writeFileSync, renameSync } from "fs"
import { homedir } from "os"
import { join } from "path"
import type { StoreState } from "./detect"

export type Cursor = Record<string, StoreState>

function stateDir(): string {
  const base = process.env.PARLAY_STATE_HOME || join(homedir(), ".parlay")
  return join(base, "robots-watch")
}
function cursorPath(): string {
  return join(stateDir(), "cursor.json")
}

export function readCursor(): Cursor {
  const p = cursorPath()
  if (!existsSync(p)) return {}
  try {
    const parsed = JSON.parse(readFileSync(p, "utf8"))
    return parsed && typeof parsed === "object" ? parsed : {}
  } catch {
    // A corrupt cursor is treated as empty → every store re-seeds (fires nothing),
    // which is the safe failure: we never replay history, we just lose one diff.
    return {}
  }
}

export function writeCursor(cursor: Cursor): void {
  const dir = stateDir()
  mkdirSync(dir, { recursive: true })
  const tmp = join(dir, `.cursor.${process.pid}.tmp`)
  writeFileSync(tmp, JSON.stringify(cursor, null, 2) + "\n")
  renameSync(tmp, cursorPath()) // atomic swap
}

File: tools/cli/internal/robotswatch/cursor.go
// Persisted per-bead status cursor for the poll-daemon: the "what we last
// saw" half of the diff. { "<store>": { "<bead-id>": "<status>" } }.
//
// Ported from cursor.ts. A corrupt or missing cursor file is treated as
// empty — every store re-seeds (fires nothing) rather than replaying
// history, the safe failure mode: we lose one diff, never replay stale
// transitions.
package robotswatch

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"

	"github.com/trillium/parlay/tools/cli/internal/config"
)

// Cursor is the whole persisted diff state, keyed by store name.
type Cursor map[string]StoreState

// stateDir is the shared state dir for both the poll cursor and the
// tailer's offset file. Honors $PARLAY_STATE_HOME the same as
// internal/config.StateHome (default ~/.parlay).
func stateDir() string {
	return filepath.Join(config.StateHome(), "robots-watch")
}

func cursorPath() string {
	return filepath.Join(stateDir(), "cursor.json")
}

func readCursor() Cursor {
	data, err := os.ReadFile(cursorPath())
	if err != nil {
		return Cursor{}
	}
	var cursor Cursor
	if err := json.Unmarshal(data, &cursor); err != nil || cursor == nil {
		return Cursor{}
	}
	return cursor
}

// writeCursor persists cursor via a same-directory temp file + atomic
// rename. Any failure here panics rather than swallowing the error — the
// caller (pollOnce, under runPollOnce's recover) is the one boundary that
// decides a bad pass must not kill the daemon, matching how cursor.ts's
// unguarded mkdirSync/writeFileSync exceptions would bubble to index.ts's
// outer try/catch.
func writeCursor(cursor Cursor) {
	dir := stateDir()
	if err := os.MkdirAll(dir, 0o755); err != nil {
		panic(err)
	}
	data, err := json.MarshalIndent(cursor, "", "  ")
	if err != nil {
		panic(err)
	}
	data = append(data, '\n')
	tmp := filepath.Join(dir, fmt.Sprintf(".cursor.%d.tmp", os.Getpid()))
	if err := os.WriteFile(tmp, data, 0o644); err != nil {
		panic(err)
	}
	if err := os.Rename(tmp, cursorPath()); err != nil { // atomic swap
		panic(err)
	}
}
